neurots/generate/grower.py

```diff
@@ -402,7 +402,10 @@ def _grow_trunks(self):
         we fit the raw binned 3d angle data and apply :func:`_3d_angles_grow_trunks`.
         """
 
-        if fit_3d_angles(self.input_parameters, self.input_distributions):
+        with_3d, self.input_parameters = fit_3d_angles(
```

Comment:
Better but you are still overwriting the instance's `input_parameters`, which shouldn't change throughout its life.

You make a local copy of the parameters and pass it to the function that needs it.
```python
with_3d, input_parameters = fit_3d_angles(...)
if with_3d:
    self._3d_angles_grow_trunks(input_parameters)
```
However, it is peculiar that you need to change the parameters from within the grower. A more natural flow would be to try do the fitting before initializing the grower, passing to it the final parameters along with the option of which grow trunk algorithm to use.

Reply:
the trick is that to get the fit one needs distributions, but parameters should not depend on distributions, if I understood @lidakanari  correctly. This means we either need an extra step before synthesis, or do it 'on the fly'. With this approach, we can do both: if we precompute the fit, this function will do nothing but check we can use 3d angle. In practice, we will always do that, where the fit will be done prior to running synthesis. This 'on the fly' approach is only for standalone use of neurots. 

I cannot pass the modified input_parameters to all the functions called later, it will be too much I think, no?

Reply:
This is not a matter of convenience, but rather one of correctness.

The `self.input_parameters` and `self.input_distributions` are, as their names indicate, inputs to the grower and are also used by classes that derive from the `NeuronGrower`, such as the `AstrocyteGrower` for instance. Therefore, we cannot modify them arbitrarily from inside the methods of the parent as they may result in unpredictable and hard-to-debug behavior.

As for where to add this step, I will leave this to @lidakanari to answer. I am not familiar with all the details of the algorithm to answer this correctly.

Reply:
voila!